mne/preprocessing/tests/test_csd.py

```diff
@@ -3,168 +3,142 @@
-    onset = raw.times[np.arange(50, n_epochs * n_times, n_times)]
-    raw.set_annotations(Annotations(onset=onset,
-                                    duration=np.repeat(0.1, 3),
-                                    description=np.repeat('foo', 3)))
-
-    events, event_id = events_from_annotations(raw)
-    epochs = Epochs(raw, events, event_id, tmin=-.2, tmax=1.34,
-                    preload=True, reject=None, picks=None,
-                    baseline=(None, 0), verbose=False)
-    sphere = (0., 0., 0., 0.095)
-    return raw, epochs, sphere
-
-
-def test_csd_matlab(raw_epochs_sphere):
+    data = loadmat(eeg_fname)['data']
+    coords = loadmat(coords_fname)['coords'] * 1e-3
+    csd = loadmat(csd_fname)['csd']
+    sphere = np.array((0, 0, 0, 0.08500060886258405))  # meters
+    sfreq = 256  # sampling rate
+    # swap coordinates' shape
+    pos = np.rollaxis(coords, 1)
+    # swap coordinates' positions
+    pos[:, [0]], pos[:, [1]] = pos[:, [1]], pos[:, [0]]
+    # invert first coordinate
+    pos[:, [0]] *= -1
+    dists = np.linalg.norm(pos, axis=-1)
+    assert_allclose(dists, sphere[-1], rtol=1e-2)  # close to spherical, meters
+    # assign channel names to coordinates
+    ch_names = [str(ii) for ii in range(len(pos))]
+    dig_ch_pos = dict(zip(ch_names, pos))
+    montage = make_dig_montage(ch_pos=dig_ch_pos, coord_frame='head')
+    # create info
+    info = create_info(ch_names=ch_names, sfreq=sfreq, ch_types='eeg')
+    # make Evoked object
+    evoked = EvokedArray(data=data, info=info, tmin=-1)
+    evoked.set_montage(montage)
+    return evoked, csd, sphere
+
+
+def test_csd_matlab(evoked_csd_sphere):
     """Test replication of the CSD MATLAB toolbox."""
-    raw, epochs, sphere = raw_epochs_sphere
-    n_epochs = len(epochs)
-    picks = pick_types(epochs.info, meg=False, eeg=True, eog=False,
-                       stim=False, exclude='bads')
-
-    csd_data = sio.loadmat(csd_fname)
-
-    montage = make_standard_montage('EGI_256', head_size=0.100004)
-    positions = np.array([montage.dig[pick]['r'] * 10 for pick in picks])
-    cosang = np.dot(positions, positions.T)
-    G = _calc_g(cosang)
-    assert_allclose(G, csd_data['G'], atol=1e-3)
-    H = _calc_h(cosang)
-    assert_allclose(H, csd_data['H'], atol=1e-3)
-    for i in range(n_epochs):
-        epochs_csd = compute_current_source_density(epochs, sphere=sphere)
-        assert_allclose(epochs_csd.get_data(), csd_data['X'], atol=1e-3)
-
-    # test raw
-    csd_raw = compute_current_source_density(raw, sphere=sphere)
+    evoked, csd, sphere = evoked_csd_sphere
+    evoked_csd = compute_current_source_density(evoked, sphere=sphere)
+    assert_allclose(linalg.norm(csd), 0.00177, atol=1e-5)
+    # If we don't project onto the sphere, we get 1e-12 accuracy here,
+    # but it's a bad assumption for real data!
+    # Also, we divide by (radius ** 2) to get to units of V/m², unclear
+    # why this isn't done in the upstream implementation
```

Comment:
This is the other thing I changed @alexrockhill -- the `_compute_csd` function has a `/ radius ** 2` to get the units to be a source density. It's unclear why the upstream implementation does not have this, as I think it will end up with units of `V` instead of `V/m²`. The latter seems like the right units, but again maybe I need to read the papers...

Otherwise, what you had and what `master` did were nearly equivalent.

Reply:
Based on the code snippets you pasted, it looks like perhaps you only factor in `radius` here (`head` in his code) if you want to scale away from the default SI units (except his is in uV, but I assume the input data are assumed to be in uV already). So indeed perhaps we should get rid of the `/ radius ** 2` in our code

Reply:
Leaving `radius ** 2` in the code makes sense conceptually because it's a current source density -- once you do all your work on the unit sphere, dividing by the actual radius gets you to the correct units (I think). Intuitively it makes sense because the smaller the head, the higher the current *source density* needs to be to achieve the same observed EEG voltages. This is also consistent with `master` and Jurgen's code, so I think we should keep it.